eo-runtime/src/main/java/EOorg/EOeolang/EOstring$EOslice.java

```diff
@@ -50,19 +50,44 @@ public class EOstring$EOslice extends PhDefault {
             final int start = new Param(rho, "start").strong(Long.class).intValue();
             final int length = new Param(rho, "len").strong(Long.class).intValue();
             final int end = length + start;
-            if (start < 0 || start > end || end > str.length()) {
-                return new PhWith(
-                    new EOerror(Phi.Φ), "msg",
-                    new Data.ToPhi(
-                        String.format(
-                            "Parameters are out of bound: start %d, len: %d, string length: %d",
-                            start, length, str.length()
-                        )
-                    )
+            Phi result;
+            if (start < 0) {
+                result = createError(
+                    "Start index must be greater than 0 but was %d",
+                    start
                 );
+            } else if (start > end) {
+                result = createError(
+                    "End index must be greater or equal to start but was %d < %d",
+                    end, start
+                );
+            } else if (end > str.length()) {
+                result = createError(
+                    "Start index + length must not exceed string length but was %d > %d",
+                    end, str.length()
+                );
+            } else {
+                result = new Data.ToPhi(str.substring(start, end));
             }
-            return new Data.ToPhi(str.substring(start, end));
+            return result;
         }));
     }
 
+    /**
+     * Building error.
+     * @param msg Formatted string for error message
+     * @param args Arguments for the formatted string
+     * @return φ containing error
+     */
+    private static Phi createError(String msg, Object... args) {
```

Comment:
@mximp why not just `error`? also, it's better to make all arguments `final` in all Java methods.